services/venice-controller/src/main/java/com/linkedin/venice/controller/UnusedValueSchemaCleanupService.java

```diff
@@ -0,0 +1,156 @@
+ */
+public class UnusedValueSchemaCleanupService extends AbstractVeniceService {
+  private static final Logger LOGGER = LogManager.getLogger(UnusedValueSchemaCleanupService.class);
+
+  private final ScheduledExecutorService executor =
+      Executors.newSingleThreadScheduledExecutor(new DaemonThreadFactory("UnusedValueSchemaCleanupService"));
+  private final VeniceControllerMultiClusterConfig multiClusterConfig;
+  private final VeniceParentHelixAdmin veniceParentHelixAdmin;
+  private final VeniceHelixAdmin admin;
+  private final int scheduleIntervalMinutes;
+  private boolean stop = false;
+  public final int minSchemaCountToKeep;
+
+  UnusedValueSchemaCleanupService(
+      VeniceControllerMultiClusterConfig multiClusterConfig,
+      VeniceHelixAdmin admin,
+      VeniceParentHelixAdmin parentHelixAdmin) {
+    this.multiClusterConfig = multiClusterConfig;
+    this.admin = admin;
+    this.scheduleIntervalMinutes = multiClusterConfig.getUnusedSchemaCleanupIntervalMinutes();
+    this.minSchemaCountToKeep = multiClusterConfig.getMinSchemaCountToKeep();
+    this.veniceParentHelixAdmin = parentHelixAdmin;
+  }
+
+  private Runnable getRunnableForSchemaCleanup() {
+    return () -> {
+      if (stop) {
+        return;
+      }
+      for (String clusterName: multiClusterConfig.getClusters()) {
+        boolean cleanupEnabled =
+            multiClusterConfig.getControllerConfig(clusterName).isUnusedValueSchemaCleanupServiceEnabled();
+
+        // Get all stores for current cluster
+        List<Store> stores = admin.getAllStores(clusterName);
+        for (Store store: stores) {
+          String storeName = store.getName();
+          // Remove schema only for batch stores
+          if (isHybridStore(store) || VeniceSystemStoreUtils.isSystemStore(storeName)) {
+            continue;
+          }
+          List<SchemaEntry> allSchemas =
+              new ArrayList<>(veniceParentHelixAdmin.getValueSchemas(clusterName, storeName));
+          if (allSchemas.size() < minSchemaCountToKeep) {
+            continue;
+          }
+
+          Set<Integer> inUseValueSchemaIds = veniceParentHelixAdmin.getInUseValueSchemaIds(clusterName, storeName);
+
+          // if any of the child colo is unreachable, skip deletion.
+          if (inUseValueSchemaIds.isEmpty()) {
+            LOGGER.warn("Could not find in-use value schemas for store {}", storeName);
+            continue;
+          }
+
+          Set<Integer> schemasToDelete = findSchemaIdsToDelete(
+              allSchemas,
+              store,
+              admin.getHelixVeniceClusterResources(clusterName).getSchemaRepository(),
+              inUseValueSchemaIds);
+
+          if (!schemasToDelete.isEmpty()) {
+            LOGGER.info(
+                "In cluster {}, store {} has the following unused schemas {}.",
+                clusterName,
+                storeName,
+                schemasToDelete);
+            if (cleanupEnabled) {
```

Comment:
I saw if you have already checked `cleanupEnabled` before the loop, so this is redundant, right?

Reply:
No, I removed the earlier check, so in a way it will collect data for unused schema without deletion.

Reply:
I see.